src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -132,6 +158,211 @@ internal void FinalizeDebugInfo()
+                        this.Context,
+                        ConvertToDebugPosition(absolutePosition),
+                        subProgram);
+
+                    if (this.sharedState.CurrentBlock != null && dIType != null)
+                    {
+                        // create the debug info for the local variable declaration
+                        diBuilder.InsertDeclare(
+                            storage: value.Value,
+                            varInfo: dIVar,
+                            location: dILoc,
+                            insertAtEnd: this.sharedState.CurrentBlock);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Creates a global function. If the debug flag is set, includes debug information
+        /// with the global function.
+        /// </summary>
+        /// <param name="spec">The <see cref="QsSpecialization"/> for the function.</param>
+        /// <param name="mangledName">The mangled name for the function.</param>
+        /// <param name="signature">The signature for the function.</param>
+        /// <param name="isDefinition">Whether this is a function definition (rather than a declaration).</param>
+        /// <returns>The <see cref="IrFunction"/> corresponding to the given Q# type.</returns>
+        internal IrFunction CreateGlobalFunction(QsSpecialization spec, string mangledName, IFunctionType signature, bool isDefinition)
+        {
+            if (this.DebugFlag && spec.Kind == QsSpecializationKind.QsBody)
+            {
+                // set up the DIBuilder
+                string sourcePath = spec.Source.CodeFile;
+                DebugInfoBuilder curDIBuilder = this.GetOrCreateDIBuilder(sourcePath);
+                DIFile debugFile = curDIBuilder.CreateFile(Path.ChangeExtension(sourcePath, ".c"));
+
+                // set up debug info for the function
+                IDebugType<ITypeRef, DIType>? retDebugType;
+                IDebugType<ITypeRef, DIType>?[] argDebugTypes;
+                retDebugType = this.GetDebugTypeFor(spec.Signature.ReturnType, curDIBuilder);
+                argDebugTypes = spec.Signature.ArgumentType.Resolution is ResolvedTypeKind.TupleType ts ? ts.Item.Select(t => this.GetDebugTypeFor(t, curDIBuilder)).ToArray() :
+                    new IDebugType<ITypeRef, DIType>?[] { this.GetDebugTypeFor(spec.Signature.ArgumentType, curDIBuilder) };
+                string shortName = spec.Parent.Name; // We want to show the user the short name in the debug info instead of the mangled name
+
+                // get the location info
+                QsNullable<QsLocation> debugLoc = spec.Location;
+                if (debugLoc.IsNull)
+                {
+                    return this.CreateFunctionNoDebug(mangledName, signature);
+                }
+
+                Position absolutePosition = debugLoc.Item.Offset; // Because spec is its own namespace, the position is already an absolute position
```

Comment:
I don't understand what you mean by "spec is its own namespace" (same for a similar comment below).

Reply:
For almost all of the syntax tree nodes, except for Namespaces of course, the position that is stored is relative to the position of the NamespaceElement ancestor as well as the stack of ancestor statements. However, with Specializations specifically, they are absolute positions instead of relative positions. I'm not sure why it was done this way, but I assume the position is made absolute for these because specializations are callables, which each become their own namespace. That's what I meant by the comment, but I can see why it's not clear -- I'll rewrite that